stage0/src/image_header.rs

```diff
@@ -2,155 +2,106 @@
-        return None;
-    }
-
-    // We've validated that the image range should be safe
-    let img_start = imagea.get_img_start();
-
-    let table_start = imagea.get_table_start();
-
-    // Step 2: Check that the table pointed to by this image is actually
-    // within our image range that we checked before
-    if !imagea.check_bounds(table_start) {
-        return None;
-    }
-
-    // The table is within bounds so accessing it will not cause a fault
-    let table = unsafe { &*imagea.get_cert_table() };
-
-    // our 'signature' is the letters cert. If this isn't valid the rest
-    // of the data is probably not valid either.
-    if table.signature != [0x63, 0x65, 0x72, 0x74] {
-        return None;
-    }
-
-    let key_start = &table.key as *const u32 as u32;
-
-    // validate that our key is fully programmed
-    if !imagea.check_bounds(key_start + table.key_size) {
-        return None;
-    }
-
-    let sig_addr = img_start + table.total_image_len;
-
-    if !imagea.check_bounds(sig_addr) {
-        return None;
-    }
-
-    // The address itself is valid so reading this is safe
-    let sig_size = unsafe { core::ptr::read_volatile(sig_addr as *const u32) };
-
-    // Check the signature
-    if !imagea.check_bounds(sig_addr + 4) {
-        return None;
-    }
-
-    if !imagea.check_bounds(sig_addr + 4 + sig_size) {
-        return None;
-    }
+    let img = Image(imagea);
 
-    // Check what is supposed to be the full image length
-    if !imagea.check_bounds(img_start + table.total_image_len) {
+    if !img.validate() {
         return None;
     }
 
-    // We can now say that the following are safe
-    // - Accessing the full image range
-    // - Accessing the key range
-    // - Accessing the signature range
-    Some(imagea)
-}
-
-// The careful observer will note that yes this is just the
-// start of an ARMv8m image with extra data shoved in the
-// vector table
-#[repr(C)]
-pub struct ImageHeader {
-    sp: u32,
-    pc: u32,
-    _vector_table: [u8; 24],
-    image_length: u32,
-    _image_type: u32,
-    header_offset: u32,
+    Some(img)
 }
 
-impl ImageHeader {
-    pub fn get_img_start(&self) -> u32 {
-        self as *const Self as u32
-    }
-
-    fn check_bounds(&self, address: u32) -> bool {
-        let start = self.get_img_start();
-
-        return address >= start && address < (start + self.image_length);
+impl Image {
+    fn get_img_start(&self) -> u32 {
+        self.0 as *const ImageVectors as u32
     }
 
-    fn get_table_start(&self) -> u32 {
-        self.get_img_start() + self.header_offset
+    fn get_header(&self) -> *const ImageHeader {
+        // SAFETY: This generated by the linker script which we trust
+        // Note that this is generated from _this_ image's linker script
+        // as opposed to the _image_ linker script but those two _must_
+        // be the same value!
+        let vector_size = unsafe { core::ptr::addr_of!(__vector_size) as u32 };
```

Comment:
Nit: `__vector_end` might be clearer? Though if you change the type to `()` as I noted on the linker script, the chances of confusion from the name seem _very low,_ so, up to you.

Reply:
This isn't the end of the vectors, `__vector_size` is holding the actual size of the vector table.